dialer.go

```diff
@@ -68,175 +63,172 @@ func NewDialer(opts *DialerOpts) Dialer {
 	if opts.WindowSize <= 0 {
 		opts.WindowSize = defaultWindowSize
 	}
-	if opts.MaxLiveConns <= 0 {
-		opts.MaxLiveConns = 1
+	if opts.LiveConns <= 0 {
+		opts.LiveConns = 2
 	}
-	if opts.MaxStreamsPerConn <= 0 || opts.MaxStreamsPerConn > maxID {
+	if opts.MaxStreamsPerConn == 0 || opts.MaxStreamsPerConn > maxID {
 		opts.MaxStreamsPerConn = maxID
 	}
-
-	if opts.RedialSessionInterval <= 0 {
-		opts.RedialSessionInterval = 5 * time.Second
-	}
-	log.Debugf("Initializing Dialer with   windowSize: %v   maxPadding: %v   maxLiveConns: %v  maxStreamsPerConn: %v   pingInterval: %v   cipher: %v",
+	log.Debugf("Initializing Dialer with   windowSize: %v   maxPadding: %v   liveConns: %v  maxStreamsPerConn: %v   pingInterval: %v   cipher: %v",
 		opts.WindowSize,
 		opts.MaxPadding,
-		opts.MaxLiveConns,
+		opts.LiveConns,
 		opts.MaxStreamsPerConn,
 		opts.PingInterval,
 		opts.Cipher)
-	liveSessions := make(chan sessionIntf, opts.MaxLiveConns)
-	liveSessions <- nullSession{}
-	return &dialer{
-		windowSize:            opts.WindowSize,
-		maxPadding:            opts.MaxPadding,
-		maxStreamsPerConn:     opts.MaxStreamsPerConn,
-		maxLiveConns:          opts.MaxLiveConns,
-		idleInterval:          opts.IdleInterval,
-		pingInterval:          opts.PingInterval,
-		redialSessionInterval: opts.RedialSessionInterval,
-		pool:            opts.Pool,
-		cipherCode:      opts.Cipher,
-		serverPublicKey: opts.ServerPublicKey,
-		liveSessions:    liveSessions,
-		numLive:         1, // the nullSession
-		emaRTT:          ema.NewDuration(0, 0.5),
+	var lc ClientLifecycleListener
+	if opts.Lifecycle != nil {
+		lc = opts.Lifecycle
+	} else {
+		lc = NoopClientLifecycleListener()
+	}
+	d := &dialer{
+		windowSize:        opts.WindowSize,
+		maxPadding:        opts.MaxPadding,
+		maxStreamsPerConn: opts.MaxStreamsPerConn,
+		pingInterval:      opts.PingInterval,
+		pool:              opts.Pool,
+		cipherCode:        opts.Cipher,
+		serverPublicKey:   opts.ServerPublicKey,
+		liveSessions:      make(chan sessionIntf, opts.LiveConns),
+		emaRTT:            ema.NewDuration(0, 0.5),
+		dial:              opts.Dial,
+		pendingSessions:   make(chan *pendingSession, opts.LiveConns),
+		lifecyle:          lc,
+		name:              opts.Name,
+	}
+	d.lifecyle.OnStart()
+	for i := 0; i < opts.LiveConns-1; i++ {
+		d.pendingSessions <- newPendingSession(d.name)
+	}
+
+	// We create another background session with a shorter dial timeout to ensure liveness in the case of network
+	// disruptions.
+	d.pendingSessions <- &pendingSession{
```

Comment:
Why is this not using `newPendingSession`? Also, should we make dialTimeout and sleepOnError configurable?

Reply:
we'd have to add more arguments to `newPendingSession` but certainly could do that.